app/controllers/users_controller.rb

```diff
@@ -165,4 +169,16 @@ def verify_active_session
       return
     end
   end
+
+  def count_failure(ip)
+    Rails.cache.write("failed_login_#{ip}", get_failures(ip)+1, :expires_in => 5.minutes)
+  end
+
+  def get_failures(ip)
+    Rails.cache.fetch("failed_login_#{ip}") {0}
+  end
```

Comment:
Please add a check that IP is not empty string or nil because then it would essentially test for all users. I can imagine that in some constrained environments, IP could become empty or something.

Reply:
done